gk/main.c

```diff
@@ -2135,23 +2051,157 @@ process_pkts_back(uint16_t port_back, uint16_t port_front,
+{
+	int i;
+	struct ip_flow *flow_arr[num_policies];
+	int done_lookups;
+	uint32_t flow_hash_val_arr[num_policies];
+	int32_t pos_arr[num_policies];
+
+	for (i = 0; i < num_policies; i++) {
+		flow_arr[i] = &policies[i]->policy.flow;
+		flow_hash_val_arr[i] = policies[i]->flow_hash_val;
+	}
+
+	done_lookups = 0;
+	while (done_lookups < num_policies) {
+		int ret;
+		int r;
+		uint32_t num_keys = num_policies - done_lookups;
+		if (num_keys > RTE_HASH_LOOKUP_BULK_MAX)
+			num_keys = RTE_HASH_LOOKUP_BULK_MAX;
+
+		ret = rte_hash_lookup_bulk_with_hash(
+			instance->ip_flow_hash_table,
+			(const void **)&flow_arr[done_lookups],
+			(hash_sig_t *)&flow_hash_val_arr[done_lookups],
+			num_keys, &pos_arr[done_lookups]);
+		if (ret != 0) {
+			GK_LOG(NOTICE,
+				"failed to find multiple keys in the hash table at lcore %u\n",
+				rte_lcore_id());
+		}
+
+		r = done_lookups + num_keys;
+		for (i = done_lookups; i < r; i++) {
+			int pos = pos_arr[i];
+			struct flow_entry *fe;
+			struct ggu_policy *policy = &policies[i]->policy;
+
+			if (pos >= 0)
+				fe = &instance->ip_flow_entry_table[pos];
+			else if (instance->num_scan_del > 0)
+				continue;
+			else {
+				struct gk_fib *fib = look_up_fib(
+					&gk_conf->lpm_tbl, &policy->flow);
+				if (fib == NULL || fib->action !=
+						GK_FWD_GRANTOR) {
+					/*
+					 * Drop this solicitation to add
+					 * a policy decision.
+					 */
+					char err_msg[128];
+					ret = snprintf(err_msg, sizeof(err_msg),
+						"gk: at %s initialize flow entry error",
+						__func__);
+					RTE_VERIFY(ret > 0 &&
+						ret < (int)sizeof(err_msg));
+					print_flow_err_msg(
+						&policy->flow, err_msg);
+					continue;
+				}
+
+				/*
+				 * The function add_ggu_policy_bulk() only
+				 * fills up GK_GRANTED and GK_DECLINED states.
+				 * So, it doesn't need to call
+				 * initialize_flow_entry().
+				 */
+				fe = add_new_flow_from_policy(fib,
+					policy, instance, gk_conf,
+					flow_hash_val_arr[i]);
+				if (fe == NULL)
+					continue;
+			}
+
+			update_flow_entry(fe, policy);
+		}
+
+		done_lookups += num_keys;
+	}
+}
+
 static void
 process_cmds_from_mailbox(
 	struct gk_instance *instance, struct gk_config *gk_conf)
 {
 	int i;
 	int num_cmd;
+	int num_policies = 0;
 	unsigned int mailbox_burst_size = gk_conf->mailbox_burst_size;
 	struct gk_cmd_entry *gk_cmds[mailbox_burst_size];
+	struct gk_add_policy *policies[mailbox_burst_size];
 
 	/* Load a set of commands from its mailbox ring. */
         num_cmd = mb_dequeue_burst(&instance->mb,
 		(void **)gk_cmds, mailbox_burst_size);
 
-        for (i = 0; i < num_cmd; i++) {
-		process_gk_cmd(gk_cmds[i], instance, gk_conf);
+        for (i = 0; i < num_cmd; i++)
+		process_gk_cmd(gk_cmds[i], policies, &num_policies, instance);
+
+	if (instance->num_scan_del == 0)
+		add_ggu_policy_bulk(policies, num_policies, instance, gk_conf);
+
+        for (i = 0; i < num_cmd; i++)
 		mb_free_entry(&instance->mb, gk_cmds[i]);
```

Comment:
1. Add `mb_free_entry_bulk()` to `include/gatekeeper_mailbox.h` and use it here instead of this loop.
```c
static inline void
mb_free_entry_bulk(struct mailbox *mb, void * const *obj_table, unsigned int n)
{
	rte_mempool_put_bulk(mb->pool, obj_table, n);
}
```
2. Search the code of Gatekeeper and replace direct uses of `rte_mempool_put_bulk()` with `mb_free_entry_bulk()`. This item should be a small, new patch on this pull request.